feat: enable const Container constructor

muhammadkamel opened the pull request:
Refactor [Container](cci:2://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:245:0-488:1) to use `const` constructor

This PR refactors the [Container](cci:2://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:245:0-488:1) widget to allow `const` construction.

**Changes:**
*   Converted [Container](cci:2://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:245:0-488:1) constructor to `const`.
*   Promoted `width` and `height` to `final` fields.
*   Moved constraint calculation and parameter validation from the constructor to the [build](cci:1://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:384:2-457:3) method.
*   Updated [debugFillProperties](cci:1://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:99:2-109:3) to include `width` and `height`.

**Benefits:**
*   **Performance:** Enables `const` optimizations (canonicalization), reducing memory usage and GC pressure.
*   **Build Efficiency:** Allows the widget tree to skip rebuilding subtrees when `const Container` instances are reused.

**Migration Notes:**
*   Invalid parameters (e.g. negative margins) that were previously caught in the constructor for non-const values will now be caught during the [build](cci:1://file:///Users/mohammadkamel/flutter/packages/flutter/lib/src/widgets/container.dart:384:2-457:3) phase. `const` construction will still catch some errors at compile time.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

muhammadkamel committed:
feat: enable const Container constructor by moving constraint calculations and assertions to the build method.

muhammadkamel committed:
feat: add tests for DecoratedBox debugFillProperties and Container padding interaction

muhammadkamel committed:
test: verify Container does not crash in 0x0 environment

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Address code review feedback: remove duplicate comment and misplaced ignore

muhammadkamel committed:
Remove duplicate comment for child property

Removed duplicate documentation for the child property.

muhammadkamel committed:
Simplify invalidMarginContainer definition

Refactor invalidMarginContainer declaration for clarity.

muhammadkamel committed:
Merge remote-tracking branch 'origin/refactor/container' into refactor/container

muhammadkamel committed:
Merge branch 'refactor/container' of https://github.com/muhammadkamel/flutter into refactor/container

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Merge branch 'flutter:master' into refactor/container

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Merge branch 'master' into refactor/container

Piinks commented on packages/flutter/lib/src/widgets/container.dart:
```diff
@@ -385,9 +381,17 @@ class Container extends StatelessWidget {
-    if (child == null && (constraints == null || !constraints!.isTight)) {
+    if (child == null &&
+        (constraints == null || !constraints!.isTight) &&
+        width == null &&
```
Can you resolve the constraints before here in the build method with width and height as they were previously done in the constructor? That way we don't have to carry forward constraints, width and height to validate.

Piinks commented on packages/flutter/lib/src/widgets/container.dart:
```diff
@@ -430,8 +434,14 @@ class Container extends StatelessWidget {
-    if (constraints != null) {
-      current = ConstrainedBox(constraints: constraints!, child: current);
+    // Effective constraints calculation (moved from constructor)
+    final BoxConstraints? effectiveConstraints = (width != null || height != null)
```
Bingo! Can we resolve this earlier?

Piinks commented on packages/flutter/test/widgets/container_test.dart:
```diff
@@ -8,26 +8,22 @@
 void main() {
   group('Container control tests:', () {
-    final container = Container(
+    const container = Container(
```
This might actually make this hard to land, making it a breaking change. Many many instances like this will fail analysis without const being there.

Piinks reviewed (commented):
Hey @muhammadkamel thanks for contributing! I noticed there isn't a linked issue, is there one you can provide that this change would resolve?
This could be a pretty disruptive change, can you share more about the motivation?

muhammadkamel committed:
Address review feedback for const Container constructor

- Revert unrelated changes to card.dart
- Restore original error message for color/decoration conflict
- Move effectiveConstraints calculation to the start of build() method
  to avoid carrying width/height/constraints separately through the method

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel commented:
> > I ran a small benchmark comparing the current Container implementation with a const constructor version, and the results show fewer unnecessary rebuilds when using a const constructor.
> 
> Typically in this case, we would land such a benchmark test to add it to the other performance benchmarks we have. Then once a baseline is established, we would expect to be able to tangibly measure the improvement with the change to verify it. :) Is there an existing benchmark that this change could impact? dev/benchmarks contains a lot of them, the dashboard is at https://flutter-flutter-perf.luci.app/e

Thanks for the guidance, @Piinks! I looked at the existing benchmarks and found multi_widget_construction_perf which uses Container, but the containers there use dynamic values so they can't be const.

To properly measure the benefit of const Container, I could create a new benchmark that specifically tests rebuild performance when Container instances can be canonicalized (same const Container reused across rebuilds vs creating new instances each time).

Would you prefer:

A new dedicated benchmark for const widget construction/rebuild (could be useful for other widgets too)
An extension to an existing benchmark
I can land the benchmark first to establish a baseline before the const Container change.

muhammadkamel commented:
Thanks for the suggestion, @Piinks! I've created a new benchmark `const_vs_non_const_container_perf` in `dev/benchmarks/macrobenchmarks` that specifically measures rebuild performance with const vs non-const Container widgets.

The benchmark creates a grid where half uses const Containers (canonicalized) and half uses non-const Containers (new instances each frame), allowing direct comparison of rebuild efficiency.

Would you prefer I land the benchmark in a separate PR first to establish a baseline, then follow up with the const Container change?

Piinks commented:
Thanks very much for that! Wow, appreciate the quick response. :)

I would like to give the other reviewers a chance to weigh in on this. I think the impact will still be pretty big for everyone that will need to add `const` throughout their applications, so we should be careful to consider the tradeoffs. :)

muhammadkamel committed:
Merge branch 'master' into refactor/container

muhammadkamel committed:
Refactor Container assertion tests into a group for negative margin.

muhammadkamel committed:
Merge branch 'flutter:master' into refactor/container

justinmc reviewed (commented):
Thanks for the bold proposal! I have wondered why Container isn't const myself before.

I think this comes down the following tradeoff: 1. const is better for performance, versus 2. asserts in the initializers are better for avoiding leaks and bugs.

We have leak tracking disabled in some tests because of asserts in lifecycle methods: https://github.com/flutter/flutter/issues/157470. I typically prefer to put the asserts in the initializers for the architectural benefits, despite the benefits of const. Given the fact that there would be a breaking change migration to do for this PR, I definitely feel inclined to make the same recommendation and close this PR.

If we wanted to land this change, I think we'd have to show:

 1. The performance benefits of const here are significant and give real benefits to Flutter app developers, clearly outweighing the risk of leaks or other bugs.
 2. This PR passes tests and doesn't cause any leak test failures.
 3. The migration for the lint rule breakage is doable and worth it.

muhammadkamel committed:
Add const Container performance benchmark

This benchmark measures:
- Build time comparison between const and non-const Container
- Widget rebuild skip optimization with const widgets
- Demonstrates the performance benefits of the const Container constructor

muhammadkamel committed:
Merge branch 'flutter:master' into refactor/container

dkwingsmt commented:
Since the team is leaning against this design, I'm closing this PR for now. We can definitely still discuss it if you'd like to continue pursuing, especially with proper metrics (notably the performance gain by enabling `const` here), but I'd like us to move the discussion under an issue for better tracking. In general, the best practice is always to make sure that a change is agreed upon in an issue before turning it into a PR. Regardless of the result, I really appreciate your PR and discussion!
